BCStudentSoftwareDevTeam/celts
# Rsvp Limit and Waitlist

## Resolves issue #748 


## For Issue #748:
 
Added a new column to the Event table `rsvpLimit` and created a function `getEventRsvpCountsForTerm()` in events.py to get the amount of people that have RSVPd to an event. The RSVP button is enabled/disabled based off the current RSVP amount. The wording on the Event View page now changes based off of whether or not RSVP is required, if there is a limit or not, and how many spots are left if there is a limit. When the RSVP limit is reached volunteers will be added to the Waitlist table, which is under the RSVP table on the `track_volunteers` page, and Admin can  delete and add users from the waitlist table to the RSVP table manually, regardless of limit. The waitlist table should only be visible when there are people in it. 

# Did: 
- Reordered event tabs while viewing events and added a new `RSVP History` tab. 
- Added ability for Admin to create a RSVP limit for events.
- Added a `rsvpLimit `column to the Event model and a `rsvpWaitlist `column to the EventRsvp table. 
- Added a new waitlist table to the `track_volunteers` page. The waitlist table should only appear when there is a person in it. 
- Created a `Add to RSVP` button in the waitlist table which allows CELTS Admin to move someone from the waitlist to RSVP regardless of the current number of volunteers that have RSVPd. This is only manual and volunteers are never moved from waitlist to RSVP as was requested by CELTS. 
- Made the `Add Volunteer` modal name and button change to accurately describe what they are doing. 
- Made sure all the `RSVP` buttons in the application now change to `Join Waitlist` if the RSVP limit has been reached. 
- Added a status message on the `track_volunteers` page when the RSVP limit has been reached to let the user know the people they are adding will be put into the waitlist. 
- Added wording the the `View Event` page so that potential volunteers will be aware when they are being put on a waitlist. 
- Created a new `EventRsvpLogs` model that will have the RSVP history in it. 

# To Test:
- Checkout the branch `rsvpAndVolunteerRefactor`
- Reset your database with `database/reset_database.sh test`
- Run the test suite with `tests/run_tests.sh ` and verify that all tests pass
- Run the application with `flask run` 
- Create an event with a limit 
- Add volunteers through the `Add Volunteers` modal on the `track_volunteers` page. Once you reach the limit you have set verify that the warning about future additions being added to the waitlist appear. 
- Verify that the waitlist table does not appear until there is someone in it. Also verify that the table disappears when you delete someone from the waitlist or you move them to RSVP. 
- Verify that the `Add to RSVP` works as it should.
- Check the wording on `Event View` and make sure it says the limit when there is one, informs the user they will be added to a waitlist when the limit has been reached, and just says RSVP required when there is no limit. 
- Verify the title and button in the `Add Volunteer` modal change to accurately reflect what they are doing as you add people to the RSVP and waitlist tables. 
- Verify that all of the `RSVP` buttons change to `Join Waitlist` when the RSVP limit has been reached.
